design-documents/scalars-for-templates.md

```diff
@@ -0,0 +1,194 @@
+because there are no filters provided in the template. _'escape'_ filter will be applied.
+ 
+Next _Template_ will try to find _'trans'_ directive, it will pass _('Hello, %name! %text',
+null, ['name' => 'Karen Smith', 'text' => 'txt])_ to it's _process()_ method. The directive's
+_getDefaultFilters()_ will not be called since _'raw'_ filter is specified.
+ 
+While this SPI does not allow introducing all kinds of directives like your own _switch-case_ directive or a _for_ loop
+directive it is definitely enough to cover 3rd-party developers needs.
+ 
+SPI will be created for us to be able to represent more complex directives with their own syntax like `if` and
+`for` directives:
+```php
+interface DirectiveProcessorInterface
+{
+    public function process(array $construction, Template $template): string;
+    
+    public function getRegExp(): string;
+}
+```
+_Template_ will call these directives one by one, use regular expression they provide and call their `process` method.
+By having access to raw text used inside a template these directives will be able to employ their own syntax and provide
+more complex behavior.
+ 
+In fact there will be a directive implementing this interface which will use general syntax
+`{{directive value param1=val1|filter}}inner HTML{{/directive}}` parse it and give it to classes implementing
+_SimpleDirectiveProcessorInterface_. This way developers who only need to implement a basic directive will use the
+simplified interface and other will be able to employ more low-level SPI.
+ 
+Another thing that 3rd party developers might want to extend/add are filters, I propose such SPI:
+```php
+interface FilterProcessorInterface
+{
+    /**
+     * Process a value.
+     *
+     * @param string $value HTML returned by a directive.
+     * @return string
+     */
+    public function filterValue(string $value): string;
+    
+    /**
+     * This filter's unique name.
+     *
+     * @return string
+     */
+    public function getName(): string;
+}
+```
+ 
+_Template_ class would receive a list of such filters and use them accordingly.
+ 
+###### Old directive processors
+Currently directives are being processed by methods of _Template_ class following _\<Directive Name\>Directive()_ pattern.
+This forces developers to create child classes of _Template_ in order to introduce new directives or modify existing ones.
+This way of working with directives has to be deprecated but preserved in order to keep backward compatibility.
+In the next minor version we will rewrite all existing directives using
+_DirectiveProcessorInterface_ and _SimpleDirectiveProcessorInterface_. In the meantime only new directives will be
+created using this SPI.
+ 
+We may create a dev-docs page explaining developers how to introduce directives and provide examples by creating new
+directives in order to replace _\$this.method()\_ usages inside E-mail templates (like using $this.getUrl() method).
+
+#### Backward compatibility
```

Comment:
What is the background behind this proposal? We need to have strong practical reasoning to introduce such changes. Have anyone been affected by described vulnerabilities in the past, please provide references.